### PR TITLE
Exception mapper betterments

### PR DESCRIPTION
**Pull request by kpavlov**
## Issue
Improve the ExceptionMapper extensibility and handling timeouts.
This is required for #2990 

## Change
- **Breaking change: SocketTimeoutException root cause is now translated to TimeoutException**
- Added a utility method 'with' to apply consumer actions on objects, enabling inline object modifications in a functional style. Included comprehensive test coverage to ensure correctness, handling null and non-null cases.
- Mark `repeat` method as deprecated and refactor it to use `String.repeat`. Add `@Nullable` annotations to clarify nullability in the `with` method. Suppress warnings for HTTP URL usage in the `readBytes` method for cleaner code.
- Introduced `findRoot` and `findCause` methods to traverse the causal chain of throwables. These utilities help identify the root cause or specific cause type of an exception, enhancing error-handling capabilities.
- Introduced a new constructor in the HttpException class that allows wrapping a throwable and a status code.
- Introduced comprehensive unit tests for ExceptionMapper to validate exception mapping behavior.
- Enhanced DefaultExceptionMapper to support custom and default HTTP status code extraction, improving flexibility and handling of various edge cases.

## General checklist

- [ ] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Exceptions.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package dev.langchain4j.internal;
 
 import dev.langchain4j.Internal;
+import org.jspecify.annotations.Nullable;
```
Comment (mariofusco) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/Exceptions.java:4]: I know that Intellij (annoyingly) insert these kind of annotations when refactoring code, but please let's avoid them.

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Utils.java [hunk 2]
```diff
@@ -19,32 +23,30 @@
...
+import java.util.function.Consumer;
 import java.util.function.Supplier;
-
-import static java.net.HttpURLConnection.HTTP_OK;
-import static java.nio.charset.StandardCharsets.UTF_8;
-import static java.util.Collections.unmodifiableList;
-import static java.util.Collections.unmodifiableMap;
-import static java.util.Collections.unmodifiableSet;
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
```
Comment (mariofusco) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/Utils.java:29]: Same as above.

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Exceptions.java (hunk 1 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/Exceptions.java:4]: I don't agree, @mariofusco. This is not a noise, but a workaround for missing null-safety in Java. Without this annotations Intyellij and Kotlin compiler can't make any reasoning about field and method result nullability. It actually pays off very well by highlighting possible NPEs and failing Kotlin compilation when NPE will be thrown.
I recommend annotating all Java code instead, especially for public APIs, also for documentation reasons.

https://jspecify.dev/docs/whether/

File: langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java [hunk 3]
```diff
@@ -36,16 +40,35 @@ default <T> T withExceptionMapper(Callable<T> action) {
...
+            final var statusCode = httpStatusCodeExtractor.apply(e);
+            if (statusCode != null) {
+                return mapHttpStatusCode(e, statusCode);
             }
 
-            if (rootCause instanceof UnresolvedAddressException) {
+            Throwable rootCause = findRoot(e);
+            if (!(e instanceof UnresolvedModelServerException) && rootCause instanceof UnresolvedAddressException) {
                 return new UnresolvedModelServerException(rootCause);
+            } else if (!(e instanceof TimeoutException) && rootCause instanceof SocketTimeoutException) {
```
Comment (dliubarskyi) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java]: why is `!(e instanceof TimeoutException)` check required?

File: langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java [hunk 4]
```diff
@@ -36,16 +40,35 @@ default <T> T withExceptionMapper(Callable<T> action) {
 
     class DefaultExceptionMapper implements ExceptionMapper {
 
+        private final Function<Throwable, Integer> httpStatusCodeExtractor;
+
+        public DefaultExceptionMapper(Function<Throwable, Integer> httpStatusCodeExtractor) {
+            this.httpStatusCodeExtractor = httpStatusCodeExtractor;
```
Comment (dliubarskyi) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java:46]: Why is this needed? Can't find any use of a custom extractor in this PR or in [Gemini PR](https://github.com/langchain4j/langchain4j/pull/2990/files)

File: langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java (hunk 3 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java]: I wanted to avoid loops, but might be it's better to check if TimeoutExceptuon is somewhere in the cause chain

File: langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java (hunk 4 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/ExceptionMapper.java:46]: Right now it's not needed. 

I ~~was lazy and~~ wrapped custom Gemini http exception with lc4j HttpException, so default mapping works. 

Not sure if it's a better approach, than creating specific lc4j exceptions directly, without HttpException in the cause chain. For later approach, the custom status code extractor would be needed.
@dliubarskyi , do we want all client-native status code exceptions wrapped into HttpException or not?
